akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,169 @@
+        override def onPush(): Unit = {
+          //just leave it in the port
+          println("initialized: pushed")
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onPull(): Unit = pull(in)
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            pr.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            val initialized = new Initialized(flow)
+            setHandlers(in, out, initialized)
+            setKeepGoing(false)
+        }
+      }
+      class Initialized(fl: Flow[In, Out, M]) extends InHandler with OutHandler {
+        val subSource = new SubSourceOutlet[In](s"${FutureFlow.this}.subIn")
+        val subSink = new SubSinkInlet[Out](s"${FutureFlow.this}.subOut")
+
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+            }
+          }
+        }
+        subSink.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+        pr.complete {
+          Try {
+            Source.fromGraph(subSource.source).viaMat(fl)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)
+          }
+        }
+        pr.future.value.get match {
+          case Success(_) =>
+            if (isAvailable(in)) {
+              subSource.push(grab(in))
+            }
+            initializing.upstreamFailure match {
+              case OptionVal.Some(ex) =>
+                subSource.fail(ex)
+              case OptionVal.None =>
+                if (isClosed(in))
+                  subSource.complete()
+            }
+            initializing.downstreamCause match {
+              case OptionVal.Some(cause) =>
+                subSink.cancel(cause)
+              case OptionVal.None =>
+                if (hasBeenPulled(in)) {
```

Comment:
It should be `if (isAvailable(out))`, i.e. it has been pulled from out before materialization.